### PR TITLE
Test automation framework

### PR DESCRIPTION
**Pull request by NadzeyaKiz**
In TestAutomation.Epam.Tests there are tests from SeleniumAdvanced. I refactored them with my TAF.

### CODE REVIEW

File: code/.runsettings [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (MariaEroshevich) [inline code/.runsettings:1]: Add name for runsettings.

File: code/TestAutomation.Core/Browser/DriverFactory.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                }
+            }
+            catch (Exception ex)
+            {
+                // todo: add logger later MyLogger.Error($"Failed to create browser instance, {ex.Message}");
+                throw new Exception("Failed to create browser instance.", ex);
+            }
+        }
+
+        private static IWebDriver GetChromeConfiguredBrowser()
```
Comment (MariaEroshevich) [inline code/TestAutomation.Core/Browser/DriverFactory.cs:42]: It is better to create separate classes for all browsers.

File: code/TestAutomation.Core/Browser/WebDriverExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+using TestAutomation.Core.Elements;
+using TestAutomation.Utilities;
+
+namespace TestAutomation.Core.Browser
+{
+    public static class WebDriverExtensions
+    {
+        public static TimeSpan DefaultTimeout = TimeSpan.FromMilliseconds(15000);
+        public static TimeSpan DefaultSleepTimeout = TimeSpan.FromMilliseconds(1000);
+        public static string GetUrl(this IWebDriver webDriver)
```
Comment (MariaEroshevich) [inline code/TestAutomation.Core/Browser/WebDriverExtensions.cs:14]: add spaces

File: code/TestAutomation.Core/Browser/WebDriverExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+            webDriver.Navigate().GoToUrl(url);
+            return webDriver;
+        }
+
+        public static void ScrollToElement(this IWebDriver webDriver, IWebElement element)
+        {            
+            Actions actions = webDriver.GetActions();
+            actions.ScrollToElement(element).Build().Perform();
+        }
+
```
Comment (MariaEroshevich) [inline code/TestAutomation.Core/Browser/WebDriverExtensions.cs]: remove space.

File: code/TestAutomation.Core/Browser/WebDriverExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        {
+            Logger.Info("Maximize Browser");
+            webDriver.Manage().Window.Maximize();
+        }
+        public static void ScrollTop(this IWebDriver webDriver)
+        {
+            Actions actions = webDriver.GetActions();
+            actions.SendKeys(Keys.Home).Build().Perform();
+        }
+
```
Comment (MariaEroshevich) [inline code/TestAutomation.Core/Browser/WebDriverExtensions.cs]: remove space

File: code/TestAutomation.Core/Browser/WebDriverExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+                return ((IJavaScriptExecutor)webDriver!).ExecuteScript(script, args);
+            }
+            catch (WebDriverTimeoutException e)
+            {
+                Logger.Info($"Error: Timeout Exception thrown while running JS Script:{e.Message}-{e.StackTrace}");
+            }
+
+            return null!;
+        }
+        public static Actions GetActions(this IWebDriver webDriver)
```
Comment (MariaEroshevich) [inline code/TestAutomation.Core/Browser/WebDriverExtensions.cs:78]: add space

File: code/TestAutomation.Core/Browser/WebDriverExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+                Logger.Error($"Unable to Quit the browser. Reason: {ex.Message}");
+            }
+        }
+
+        public static void SwitchToWindow(this IWebDriver webDriver, string windowHandle)
+        {
+            webDriver.SwitchTo().Window(windowHandle);
+        }
+
+        public static void ClickWithFindElement(this IWebDriver driver, By by)
```
Comment (MariaEroshevich) [inline code/TestAutomation.Core/Browser/WebDriverExtensions.cs:113]: I don't think that all methods that related to working with WebElement should be here.

File: code/TestAutomation.Core/Browser/WebDriverExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+            {
+                Logger.Info($"Element {by.ToString()} is not displayed. Waiting to be displayed to become clickable");
+                driver.WaitForElementToBeVisible(element);
+            }
+            element.ClickUsingJS(driver);
+        }
+
+
+        #region Waiters
+        public static void WaitForCondition(this IWebDriver driver, Func<IWebDriver, bool> condition, TimeSpan timeout)
```
Comment (MariaEroshevich) [inline code/TestAutomation.Core/Browser/WebDriverExtensions.cs:185]: Move waiters to separate class.

File: code/TestAutomation.Core/Elements/WebElementExtension.cs [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        public static bool IsElementDisplayedOnPage(this IWebElement element, IWebElement epamLogo)
+        {
+            return element.Displayed;
+        }
+
+        public static bool IsElementEnabled(this IWebElement element)
+        {
+            return element.Enabled;
+        }
+
```
Comment (MariaEroshevich) [inline code/TestAutomation.Core/Elements/WebElementExtension.cs]: remove spaces.

File: code/TestAutomation.Core/Utilities/WebDriverCookiesExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
+﻿using OpenQA.Selenium;
+using TestAutomation.Core.Browser;
+namespace TestAutomation.Core.Utilities
+{
+    public static class WebDriverCookiesExtensions
+    {
+        public static void SetUpCookies(this IWebDriver driver, By locator)
```
Comment (MariaEroshevich) [inline code/TestAutomation.Core/Utilities/WebDriverCookiesExtensions.cs:7]: This class is not related to Core layer. Move it to PageObject layer.

File: code/TestAutomation.Core/Browser/WebDriverExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+using OpenQA.Selenium.Interactions;
+using OpenQA.Selenium.Support.UI;
+using TestAutomation.Core.Elements;
+using TestAutomation.Utilities;
+
+namespace TestAutomation.Core.Browser
+{
+    public static class WebDriverExtensions
+    {
+        public static TimeSpan DefaultTimeout = TimeSpan.FromMilliseconds(15000);
```
Comment (MariaEroshevich) [inline code/TestAutomation.Core/Browser/WebDriverExtensions.cs:11]: I think that it is better to move values for timeouts to runsettings.

File: code/TestAutomation.Core/Browser/WebDriverExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,223 @@
+﻿using OpenQA.Selenium;
+using OpenQA.Selenium.Interactions;
+using OpenQA.Selenium.Support.UI;
+using TestAutomation.Core.Elements;
+using TestAutomation.Utilities;
+
+namespace TestAutomation.Core.Browser
```
Comment (MariaEroshevich) [inline code/TestAutomation.Core/Browser/WebDriverExtensions.cs:7]: Implement Singleton pattern to guarantee that only one instance of driver will be created.

**Review by MariaEroshevich — CHANGES_REQUESTED** [review]

File: code/TestAutomation.Core/Browser/WebDriverExtensions.cs (hunk 3 above)
Comment (NadzeyaKiz) [reply code/TestAutomation.Core/Browser/WebDriverExtensions.cs:14]: done

Comment (NadzeyaKiz) [reply code/TestAutomation.Core/Browser/WebDriverExtensions.cs:14]: done

File: code/TestAutomation.Core/Browser/WebDriverExtensions.cs (hunk 4 above)
Comment (NadzeyaKiz) [reply code/TestAutomation.Core/Browser/WebDriverExtensions.cs]: done

File: code/TestAutomation.Core/Browser/WebDriverExtensions.cs (hunk 5 above)
Comment (NadzeyaKiz) [reply code/TestAutomation.Core/Browser/WebDriverExtensions.cs]: done

File: code/TestAutomation.Core/Elements/WebElementExtension.cs (hunk 9 above)
Comment (NadzeyaKiz) [reply code/TestAutomation.Core/Elements/WebElementExtension.cs]: done

File: code/TestAutomation.Core/Browser/WebDriverExtensions.cs (hunk 8 above)
Comment (NadzeyaKiz) [reply code/TestAutomation.Core/Browser/WebDriverExtensions.cs:185]: done

**Pull request merged**